src/common/quota/util.go

```diff
@@ -0,0 +1,89 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package quota
+
+import (
+	"errors"
+
+	"github.com/astaxie/beego/orm"
+	"github.com/goharbor/harbor/src/common/models"
+)
+
+const (
+	// UNLIMITED unlimited quota value
+	UNLIMITED = int64(-1)
+)
+
+type operate func(current int64, value int64) int64
+
+func reserve(current int64, value int64) int64 {
+	return current + value
+}
+
+func unreserve(current int64, value int64) int64 {
+	return current - value
+}
+
+func calcUsed(quota *models.Quota, usage *models.QuotaUsage, res Resource, op operate) (models.QuotaUsed, error) {
+	hard, err := quota.GetHard()
+	if err != nil {
+		return nil, err
+	}
+
+	used, err := usage.GetUsed()
+	if err != nil {
+		return nil, err
+	}
+
+	ret := used.Copy()
+	for key, value := range res {
+		current := op(used[key], value)
+
+		if err := isSafe(hard[key], current); err != nil {
+			return nil, err
+		}
+
+		ret[key] = current
+	}
+
+	return ret, nil
+}
+
+func isSafe(hard int64, used int64) error {
+	if used < 0 {
+		return errors.New("bad used value")
+	}
+
+	if hard == UNLIMITED {
+		return nil
+	}
+
+	if used > hard {
+		return errors.New("over the quota")
+	}
+
+	return nil
+}
+
+func withTransaction(fn func(o orm.Ormer) error) error {
```

Comment:
move it to base of DAO, it could be reused by other API.

Reply:
moved